src/main/kotlin/com/hibob/academy/types/StoreService.kt

```diff
@@ -0,0 +1,80 @@
+package com.hibob.academy.types
+
+import org.springframework.stereotype.Component
+import java.time.LocalDate
+import Cart
+import Payment
+import Check
+import Statuses
+
+const val ZERO_TOTAL_COSTS_ON_FAILED_BUY = 0.0
+const val VISA_LEGIT_NUM_LENGTH = 10
+
+fun fail(message: String) : Nothing?{
+    throw IllegalStateException(message)
+}
+
+@Component
+class StoreService {
+
+    fun pay(cart: List<Cart>, payment: Payment): Map<String, Check> {
+        val result = mutableMapOf<String, Check>()
+
+        for (c in cart) {
+            val check = checkout(c, payment)
+
+            result[c.clientId] = check
+        }
+        return result
+    }
+
+    private fun isCustomTrue(custom: Any): Boolean {
+
+        return when(custom) {
+            is Boolean -> custom
+            else -> false
+        }
+    }
+
+    private fun checkout(cart: Cart, payment: Payment): Check {
+        var totalPricesOfLegitProducts = 0.0
+
+        cart.products.forEach { product ->
+            if (isCustomTrue(product.custom)) {
+                totalPricesOfLegitProducts += product.price
+            }
+        }
+
+        val status: Statuses = when(payment) {
+            is Payment.Cash -> {
+                fail("payment is Cash")
+                Statuses.FAILURE
+                return Check(cart.clientId, Statuses.FAILURE, ZERO_TOTAL_COSTS_ON_FAILED_BUY)
+            }
+
+            is Payment.CreditCard -> {
+                if (payment.expiryDate < LocalDate.now()) {
+                    // fail("expiry date is in the past")
+                    // update status to fail
+                    Statuses.FAILURE
+                } else if (payment.limit < totalPricesOfLegitProducts) {
+                    Statuses.FAILURE
+                } else if (payment.type !in listOf(CreditCardType.VISA, CreditCardType.MASTERCARD)) {
+                    Statuses.FAILURE
+                } else if (payment.number.length != VISA_LEGIT_NUM_LENGTH) {
+                    Statuses.FAILURE
+                } else { Statuses.SUCCESS }
```

Comment:
this condition is not clear first I suggest you in case of CreditCard to call function called validateCreditCard.
then I would suggest you to split the condition to parts, and each part to use function that going to be more readable and easier to fix.

take as an example:

```
    fun checkPayment (payment: Payment ,price : Double): Boolean {
        return when(payment){
            is Payment.Cash-> fail("you cant pay with cash")
            is Payment.PayPal-> {
                     if (!payment.email.contains('@')) {
                                        Statuses.FAILURE
                                    } else {
                                        Statuses.SUCCESS
                                    }
                    }
            is Payment.CreditCard-> validateCreditCard(payment, price)
        }
    }

    private fun validateCreditCard(payment: Payment.CreditCard, price: Double) =
        isSupportedType(payment.type) && isCardDetailsValid(payment) && isNotOverTheLimit(price, payment)

    private fun isSupportedType(cardType: CreditCardType) =
        (cardType == CreditCardType.MASTERCARD || cardType == CreditCardType.VISA)

    private fun isCardDetailsValid(payment: Payment.CreditCard) =
        payment.expiryDate.isAfter(LocalDate.now())  && payment.number.length == 10

    private fun isNotOverTheLimit(price: Double, payment: Payment.CreditCard) = price < payment.limit
```
look at the condition now, I think its easier to read it

Reply:
using function to organize the code is a good point, thanks! about the  private methods, should that be inside new class only for those private function? and would you still say it's needed with the current fix? (way shorter code)

Reply:
why new class? same store service but private methods.